### ISSUE TITLE
[windows] Fluent Bit hangs with high cpu usage on Windows Server

### ISSUE THREAD

**Issue by theggelund** [fluent/fluent-bit#2185]
## Bug Report

**Describe the bug**
After running fluentbit a while it will stop sending data and hang with high cpu usage.

This might be related to #1732 

**To Reproduce**
Have lots of log files in a folder and let fluentbit read these and send them to fluentd.
The process involves parsing the logs using custom regex'es.

**Expected behavior**
Should not hang

Process dump => [Link to dump file uploaded to onedrive](https://365dips-my.sharepoint.com/:u:/g/personal/the_dips_no/ERWFck_g_UFJtApTJzlmS0wBhuSVN5dQapACZfu77BXymA?e=jafux4) The file was to big for uploading to github

**Screenshots**
Using a lot more memory than configured and uses almost all of one CPU
![fluentbit_high_cpu_and_memory](https://user-images.githubusercontent.com/648714/82301249-417d8a00-99b8-11ea-8ef2-6a0bdaf7e2c9.png)

IO stats. They are not changing so the process appears to be hanging
![fluentbit_io](https://user-images.githubusercontent.com/648714/82301257-45111100-99b8-11ea-9256-997f4638e3a8.png)

Stacktrace 1
![fluentbit_stacktrace_1](https://user-images.githubusercontent.com/648714/82301268-48a49800-99b8-11ea-997e-c29d4a0a7911.png)

Stacktrace 2
![fluentbit_stacktrace_2](https://user-images.githubusercontent.com/648714/82301272-49d5c500-99b8-11ea-8fe8-bde7b4083bb2.png)

Was running fluentbit from a console window when I was reproducing the issue, this is the output
![image](https://user-images.githubusercontent.com/648714/82301559-b781f100-99b8-11ea-9ab4-a815b5e42214.png)

After doing a Ctrl+C it gracefully ends
![image](https://user-images.githubusercontent.com/648714/82302378-cf0da980-99b9-11ea-8e26-411bc84f2962.png)

**Your Environment**
![image](https://user-images.githubusercontent.com/648714/82301435-8d303380-99b8-11ea-8652-08abea753a01.png)

* Version used:
Fluent Bit v1.4.4
* Configuration
- [-S output](https://github.com/fluent/fluent-bit/files/4649036/output.txt)
- [fluentbit_conf.zip](https://github.com/fluent/fluent-bit/files/4649078/fluentbit_conf.zip)

**Additional context**
We are trying to parse logs from a Windows server and send them via a fluentd instance to Splunk / Elastic Search. The logs are application and IIS logs

**Comment by fujimotos** [comment]
@theggelund  Thank you for the detailed report. I'll try to reproduce
the issue on my dev machine.

I'm wondering which plugin is causing the issue. Do you have some
more clue about it? (e.g. it occurs only when filter plugins are enabled)

**Comment by theggelund** [comment]
Not yet, I've been trying to reproduce the bug with less configuration but I haven't been able to yet. What I've seen is that by changing Buffer_Chunk_Size, Buffer_Max_Size & Mem_Buf_Limit to a larger value, say 20M or 30M, then the hang occurs faster.

**Comment by theggelund** [comment]
I was able to reproduce the issue with a smaller setup. I just zipped the folder for you, with the log, conf and the exe I was using. 

[sender.zip](https://github.com/fluent/fluent-bit/files/4655592/sender.zip)

The receiver is a fluentbit running on the same machine 
`.\fluent-bit.exe -i forward -p Buffer_Max_Size=1M -p Buffer_Chunk_Size=1M -o file -p Path=output.log`

Hope this helps

**Comment by theggelund** [comment]
I have tested with tail input and forward, tcp & http output. They all fail. If I use a file output it doesn't so I think it could be something with the connection / tcp.

I've also tested to see if it could be the parser but I got the same error without any line parsing so it's ok.

The final setup I tried was to use http as output but not have a http server running to receive and then I got an interesting result. Fluentbit will start reading the log file and read until the mem buffer is full and then I get a few attempts to connect and a few retries and then it fails. I've attached the log

[fluentbit_with_http_output_no_receiver.txt](https://github.com/fluent/fluent-bit/files/4656073/fluentbit_with_http_output_no_receiver.txt)

and the config
````
[SERVICE]
    Log_Level debug
    Parsers_File parsers.conf

[INPUT]
    Name    tail
    Tag     format.afile
    Path    test.log 
    Buffer_Chunk_Size   512k
    Buffer_Max_Size     1M    
    Mem_Buf_Limit       50M

[OUTPUT]
    Name    http
    Port    45000
    Format  json
```

**Comment by theggelund** [comment]
Related or same error as #2162

**Comment by fujimotos** [comment]
> I was able to reproduce the issue with a smaller setup. I just zipped the folder for you, with the log, conf and the exe I was using.

@theggelund  I spent today debugging this issue using your test case.
I think I found the root cause of this bug.

I posted my fix to #2195. If possible, can you test the patch on your
environment? I have seen no hang with it so far.

**Comment by theggelund** [comment]
I've tested the fix and it works with the last configuration I sent, but ends up in crash with a different one. 
```
@SET SERVICELOGPATH="c:\DIPS-LOG"
@SET REFRESH_INTERVAL=60

[SERVICE]
    Log_Level debug
    Parsers_File parsers.conf

[INPUT]
    Name                tail
    Tag                 format.logging
    path                ${SERVICELOGPATH}\*main.log
    Refresh_Interval    ${REFRESH_INTERVAL}

    Multiline           On
    Parser_Firstline    dips-log4net
    Parser_N            dips-log4net-multiline     

[OUTPUT]
    Name    forward
```
Output
```
PS C:\Temp\test\sender> .\fluent-bit.exe -c .\fluent-bit.conf
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

PS C:\Temp\test\sender>
```

From event viewer
```
Faulting application name: fluent-bit.exe, version: 0.0.0.0, time stamp: 0x5ec7d3c1
Faulting module name: fluent-bit.exe, version: 0.0.0.0, time stamp: 0x5ec7d3c1
Exception code: 0xc0000409
Fault offset: 0x000000000012bee8
Faulting process id: 0x1e4c
Faulting application start time: 0x01d6326c23ff15d7
Faulting application path: C:\DIPS\LogForwarder\bin\fluent-bit.exe
Faulting module path: C:\DIPS\LogForwarder\bin\fluent-bit.exe
Report Id: fe035d43-123b-431b-822c-bf62da5f1214
Faulting package full name: 
Faulting package-relative application ID: 
```

Windows Error Report 
[Report.zip](https://github.com/fluent/fluent-bit/files/4676141/Report.zip)

**Comment by theggelund** [comment]
After testing a bit more I found that the error occurs when adding `Types` parameter to regex parser. This worked previously so this is a new bug. I think that this issue could be closed and a new one must be opened with the `Types` issue.

This will fail 
```
[PARSER]
    Name    dips-log4net-multiline
    Format  regex 
    Regex   ^(?<test>.*)$
    Types test:integer
```

This works
```
[PARSER]
    Name    dips-log4net-multiline
    Format  regex 
    Regex   ^.*$
```

**Comment by fujimotos** [comment]
@theggelund I'll take a look at that issue tomorrow. Wait for me.

**Comment by fujimotos** [comment]
> After testing a bit more I found that the error occurs when adding Types parameter to regex parser. 

@theggelund I posted a fix for that issue to #2200. Combined with
#2195, I believe your issues are all resolved now.

**Comment by theggelund** [comment]
Nice. I've testet it in my environment and now it works. Thanks.

When can we expect this to end up in a release version? Will there be a 1.4.5 with keepalive and infinite loop fix or do I have to wait for a 1.5?

**Comment by fujimotos** [comment]
@eduardo I'm thinking that probably the best option here is to merge #2195
and backport it to v1.4. It should work as an immediate fix for the infinite
loop bug.

As to the root fix we have talked, it will go to v1.5 on top of #2195, as it
requires surgery on the core implementation in monkey.

**Comment by fujimotos** [comment]
@edsiper  I'm thinking that probably the best option here is to merge #2195
and backport it to v1.4. It should work as an immediate fix for the infinite
loop bug.

As to the root fix we have talked, it will go to v1.5 on top of #2195, as it
requires surgery on the core implementation in monkey.

**Comment by edsiper** [comment]
Fixed in v1.4.6: https://fluentbit.io/announcements/v1.4.6/
